contrib/test.bzl

```diff
@@ -0,0 +1,144 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A wrapper around container structure tests for Bazel.
+
+This rule feeds a built image and a set of config files
+to the container structure test framework."
+"""
+
+# The release of the container-structure-test repository to use.
+STRUCTURE_TEST_RELEASE = "v0.1.1"
+
+load(
+    "//container:bundle.bzl",
+    "container_bundle",
+)
+
+def repositories():
+    if "structure_test" not in native.existing_rules().keys():
+        native.git_repository(
+            name = "structure_test",
+            remote = "https://github.com/GoogleCloudPlatform/container-structure-test.git",
+            tag = STRUCTURE_TEST_RELEASE,
+        )
+
+def _impl(ctx):
+    config_str = ' '.join(['$(pwd)/' + c.short_path for c in ctx.files.configs])
+
+    if ctx.attr.driver == 'tar':
+        # no need to load if we're using raw tar
+        load_statement = ''
+        image_name = '$(pwd)/' + ctx.file.image_tar.short_path
+
+    else:
+        # Since we're always bundling/renaming the image in the macro, this is valid.
+        load_statement = 'docker load -i %s' % ctx.file.image_tar.short_path
+        image_name = ctx.attr.image_name
+
+    # Generate a shell script to execute structure_tests with the correct flags.
+    ctx.actions.expand_template(
+        template = ctx.file._structure_test_tpl,
+        output = ctx.outputs.executable,
+        substitutions = {
+          "%{load_statement}": load_statement,
+          "%{configs}": config_str,
+          "%{workspace_name}": ctx.workspace_name,
+          "%{test_executable}": ctx.executable._structure_test.short_path,
+          "%{image}": image_name,
+          "%{driver}": ctx.attr.driver,
+        },
+        is_executable = True
+    )
+
+    return struct(runfiles = ctx.runfiles(files =
+            [
+                ctx.executable._structure_test,
+                ctx.executable.image_tar,
+                ctx.file.image_tar
+            ] +
+            ctx.attr.image_tar.files.to_list() +
+            ctx.attr.image_tar.data_runfiles.files.to_list() +
+            ctx.files.configs,
+        ),
+    )
+
+_container_test = rule(
```

Comment:
Can you add a doc string like this https://github.com/GoogleCloudPlatform/base-images-docker/blob/master/package_managers/download_pkgs.bzl#L72
This will generate documentation using skydoc.

Reply:
done